### PR TITLE
PhaseQuad integration

### LINKED ISSUE #30061: PhaseQuad as a pair
In muon analysis we should update the context to treat phasequads as a sudo pair. This should fix it not appearing correctly in other parts of the GUI.

### PR DESCRIPTION
**Pull request by AnthonyLim23**
**Description of work.**
The phasequad part of the muon GUI was not fully integrated. It was not obvious how to add it to the analysis, particularly in time domain.

**To test MA:**
1. Open muon analysis
2. Load MUSR 62260
3. Go to the phase tab
4. Try to create a ohasequad -> get a warning
5. Make a phase table, then make a phasequad
6. You should now have 2 new workspaces in the ADS (Re and Im parts)
7. Check that the new phasequads show up on the plots
8. Go to grouping add a pair
9. Unselect all of the groups
10. Go to fitting and add a fitting function
11. Do a fit
12. Go to sequential fit and do a series of fits
13. Load some new runs (62261-62263)
14. The phasequad's will have vanished from the plots
15. Do a phasequad
16. Check you have 4 new workspace (Re and Im for both 62261 and 62263)
17. Close muon analysis

**To test FDA:**
1. Open frequency domain analysis
2. Load MUSR 62265
3. Go to transform
4. Do an FFT on the phasequad data
5. Go to fitting and do a fit

Fixes #30061 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AnthonyLim23** [comment]
Some of the changes are just flake8

**Comment by AnthonyLim23** [comment]
This should be ready for review. If changing the deadtime, you will need to recalculate the phasequad. It will not automatically update.

File: scripts/Muon/GUI/Common/muon_phasequad.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+# pylint: disable=C0111
+from Muon.GUI.Common.muon_base_pair import MuonBasePair
+
+
+class MuonPhasequad(object):
+    """
+    Simple structure to store information on a phasequad.
+
+    - The name is set at initialization and after that cannot be changed.
+    - The pair has two groups associated to it, and we store only their names.
```
Comment (Matt-Cumber) [inline scripts/Muon/GUI/Common/muon_phasequad.py:16]: Is this comment right?

File: scripts/Muon/GUI/Common/phase_table_widget/phase_table_presenter.py [hunk 2]
```diff
@@ -70,49 +73,75 @@ def create_calculation_thread(self):
...
-        run = re.search('^{}([0-9, -]+)[;,_]?'.format(self.context.data_context.instrument), input_workspace).group(1)
+    def handle_phasequad_calculation_success(self):
+        self.enable_editing_notifier.notify_subscribers()
+        self.view.enable_widget()
+        self.view.disable_cancel()
 
-        directory = get_base_data_directory(self.context, run)
+        names = [self._phasequad_obj.Re.name, self._phasequad_obj.Im.name]
+        self.current_alg = None
+        pair_added = True  # if state == 2 else False
```
Comment (Matt-Cumber) [inline scripts/Muon/GUI/Common/phase_table_widget/phase_table_presenter.py:131]: I can't see how pair_added can be false here?

File: scripts/test/Muon/muon_context_test.py [hunk 3]
```diff
@@ -12,13 +12,26 @@
 
 from mantid.api import AnalysisDataService, FileFinder
 from mantid import ConfigService
-from mantid.simpleapi import CreateWorkspace
+#from mantid.simpleapi import CreateWorkspace
```
Comment (Matt-Cumber) [inline scripts/test/Muon/muon_context_test.py:15]: If the headers are commented out can we delete them?

File: scripts/test/Muon/muon_context_test.py [hunk 4]
```diff
@@ -12,13 +12,26 @@
...
 from collections import Counter
 from Muon.GUI.Common.utilities.load_utils import load_workspace_from_filename
-from Muon.GUI.Common.ADSHandler.muon_workspace_wrapper import MuonWorkspaceWrapper
+#from Muon.GUI.Common.ADSHandler.muon_workspace_wrapper import MuonWorkspaceWrapper
```
Comment (Matt-Cumber) [inline scripts/test/Muon/muon_context_test.py:18]: Same here

File: scripts/test/Muon/phase_table_widget/phase_table_presenter_test.py [hunk 5]
```diff
@@ -257,6 +200,38 @@ def test_that_enable_observer_calls_on_view_when_triggered(self):
...
+        self.context.group_pair_context.add_pair_to_selected_pairs = mock.Mock()
+        self.presenter.selected_phasequad_changed_notifier.notify_subscribers = mock.Mock()
+        self.presenter._phasequad_obj = MuonPhasequad("test", "table")
+        self.presenter.handle_phasequad_calculation_success()
+
+        self.context.group_pair_context.add_pair_to_selected_pairs.assert_any_call("test_Re_")
+        self.context.group_pair_context.add_pair_to_selected_pairs.assert_any_call("test_Im_")
+        self.presenter.selected_phasequad_changed_notifier.notify_subscribers.assert_any_call({"is_added":True, "name":"test_Re_"})
+        self.presenter.selected_phasequad_changed_notifier.notify_subscribers.assert_any_call({"is_added":True, "name":"test_Im_"})
+
```
Comment (Matt-Cumber) [inline scripts/test/Muon/phase_table_widget/phase_table_presenter_test.py:234]: Worth having a test adding phasequad with a duplicate name both with pairs and another pahesquad. Should handle the unique Real and Imaginary parts as well

**Review by Matt-Cumber — CHANGES_REQUESTED** [review]
Overall some really good changes here. A few minor comments below. I think some of the error handling could be cleared up. For example if you try add a phase table with no data you get an error message "NoneType object has no attribute detectors", would be nice if instead it said Please load some data first or something similar. Co-add is not possible but can be saved for a later date if requested by scientists. However, once you co-add and realise it is not possible to add phasequads, there is a persistent error message of "list.remove(x): x not in list" which shows up and doesn't go until restarting the interface. 

Some validation on the phase properties table for First Good Data and Last Good Data would be nice. As if you make last smaller than first and try make a table it doesn't let you, so this could be handled with a warning/error as soon as the data in the table is changed.

There is an unfriendly error message when you create phasetable twice with the only difference being output fit information is ticked. Also prints a lot of repeated warnings in the log. This causes problems trying to create a new phasetable on the same data, you need to load new data in then load back the old data to fix it.

File: scripts/Muon/GUI/Common/contexts/muon_context.py [hunk 6]
```diff
@@ -169,11 +175,34 @@ def calculate_all_pairs(self):
...
     def _calculate_pairs(self, rebin):
         for run in self._data_context.current_runs:
+
+            self._update_phasequads(rebin)
+            # construct the pairs
             for pair in self._group_pair_context.pairs:
-                pair_asymmetry_workspace = self.calculate_pair(
-                    pair, run, rebin=rebin)
+                if isinstance(pair, MuonPair):
+                    print(pair.name)
```
[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/contexts/muon_context.py:200]:
```suggestion
```

File: scripts/Muon/GUI/Common/muon_phasequad.py (hunk 1 above)
Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/muon_phasequad.py:16]: nope

Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/muon_phasequad.py:16]:
```suggestion
    - The phasquad has two pairs associated to it and are stored as MuonBasePairs.
```

File: scripts/Muon/GUI/Common/muon_phasequad.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    - The balance parameter is stored and modifiable.
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/muon_phasequad.py:17]:
```suggestion
```

File: scripts/Muon/GUI/Common/muon_phasequad.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    - The workspace associated to the pair can be set, but must be of type MuonWorkspaceWrapper.
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/muon_phasequad.py:18]:
```suggestion
    - The workspaces associated with the pairs can be set, but must be of type MuonWorkspaceWrapper.
```

File: scripts/Muon/GUI/Common/phase_table_widget/phase_table_presenter.py (hunk 2 above)
Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/phase_table_widget/phase_table_presenter.py:131]: at present it cannot. I left this in for when we have a table, with the analyse tick box.

File: scripts/test/Muon/muon_context_test.py (hunk 3 above)
Comment (AnthonyLim23) [reply scripts/test/Muon/muon_context_test.py:15]:
```suggestion
```

File: scripts/test/Muon/muon_context_test.py (hunk 4 above)
Comment (AnthonyLim23) [reply scripts/test/Muon/muon_context_test.py:18]:
```suggestion
```

File: scripts/test/Muon/phase_table_widget/phase_table_presenter_test.py (hunk 5 above)
Comment (AnthonyLim23) [reply scripts/test/Muon/phase_table_widget/phase_table_presenter_test.py:234]: thats a good idea :)

**Comment by AnthonyLim23** [comment]
I have added a check before removing the data, that should prevent the x not in list error. 

For the phase table changes I have made a new issue https://github.com/mantidproject/mantid/issues/30168
For co-add I have a new issue https://github.com/mantidproject/mantid/issues/30169

The test is a good idea, but I think would be better once we have factored out the repeated code.

**Review by Matt-Cumber — APPROVED** [review]
The code looks good and the new issues created should solve any remaining problems. Nice job 👍

**Pull request merged**